command/agent/template/template.go

```diff
@@ -247,12 +247,18 @@ func newRunnerConfig(sc *ServerConfig, templates ctconfig.TemplateConfigs) (*ctc
 		attempts = sc.AgentConfig.Vault.Retry.NumRetries
 	}
 
+	// We enable retry if it's >= 0 and it was explicitly specified in the
+	// config.
+	// Note: If NumRetries is explicitly set to 0, it technically means
+	// that the client will retry 12 times per request/attempt but this
+	// shouldn't matter since template will indefinitely retry.
+	enabled := attempts >= 0 && sc.AgentConfig.Vault.Retry.NumRetriesRaw != nil
+
 	// Use the cache if available or fallback to the Vault server values.
 	// For now we're only routing templating through the cache when persistence
 	// is enabled. The templating engine and the cache have some inconsistencies
 	// that need to be fixed for 1.7x/1.8
 	if sc.AgentConfig.Cache != nil && sc.AgentConfig.Cache.Persist != nil && len(sc.AgentConfig.Listeners) != 0 {
-		attempts = 0
```

Comment:
Don't we still want to disable retries if using persistent cache?

Reply:
We want to support the ability for template to indefinitely retry regardless of whether persistent cache is enabled. Since the cache client doesn't support indefinite retries and 0 on template really mean infinite retries, we can't hardcode 0 here.

I think that the logic that's missing here is something along the lines of:

```go
// Disable template retries if attempts is greater than zero
// so that we don't end up performing n^2 retries by both
// template and the cache client.
if attempts > 0 {
    enabled = false
}
```

Reply:
Addressed in 9fab0c1